### ISSUE TITLE
Test failure in test_memory_format_operators and faulty test

### ISSUE THREAD

**Issue by Flamefire** [pytorch/pytorch#48591]
## 🐛 Bug

I'm getting test failures in test_memory_format_operators_cpu:

```
FAIL: test_memory_format_operators_cpu (__main__.TestTorchDeviceTypeCPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/install_pt/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 273, in instantiated_test
    result = test_fn(self, *args)
  File "test_torch.py", line 15096, in test_memory_format_operators
    torch.channels_last)
  File "test_torch.py", line 15065, in _test_helper
    self.assertEqual(result, result_c)
  File "/tmp/install_pt/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 1136, in assertEqual
    self.assertTrue(result, msg=msg)
AssertionError: False is not true : Tensors failed to compare as equal! With rtol=1.3e-06 and atol=1e-05, found 320 element(s) (out of 768) whose difference(s) exceeded the margin of error (including 320 nan comparisons). The greatest difference was nan (13.80231761932373 vs. nan), which occurred at index (0, 0, 0, 4).

----------------------------------------------------------------------
Ran 1 test in 0.301s
```

I believe there is also a bug here: Some functions modify the x tensor, e.g. `x.logit_()`, so the next tests are affected. This is potentially problematic as that operations introduces quite a number of NaN values.

## To Reproduce

Either run `python test_torch.py TestTorchDeviceTypeCPU.test_memory_format_operators_cpu`

or use this code based on the test:
```
def do(x):
  x_c = x.contiguous()
  x.logit_()
  x_c.logit_()
  print("x   %s" % x)
  print("x_c %s" %x_c)
  x_c = x.contiguous()
  print("xee  %s" % x.logit(1e-6))
  print("x_c  %s" % x_c.logit(1e-6))

x = torch.randn((1,2,2,8)).contiguous(memory_format=torch.channels_last)
do(x)
```

The last 2 outputs should be equal but often are not. It seems to be a problem with the handling of NaNs as removing the in-place logit_() call avoids the issue.

## Environment

 - PyTorch Version (e.g., 1.0): 1.7.0
 - OS (e.g., Linux): REHL 7.8
 - How you installed PyTorch (`conda`, `pip`, source): source
 - Python version: 3.7.4
 - Any other relevant information: POWER architecture (can't reproduce on x86 so far)

One such output looks like this:

```
x   tensor([[[[    nan,     nan,     nan,  2.8484,  1.4755,     nan,     nan,               nan],
          [    nan,     nan,     nan,     nan, -0.5471,     nan,     nan,               nan]],

         [[    nan,     nan,     nan,     nan, -2.4072,  0.5984,     nan,               nan],
          [    nan,     nan,  0.7890,     nan,     nan,     nan,     nan,            0.3440]]]])
x_c tensor([[[[    nan,     nan,     nan,  2.8484,  1.4755,     nan,     nan,               nan],
          [    nan,     nan,     nan,     nan, -0.5471,     nan,     nan,               nan]],

         [[    nan,     nan,     nan,     nan, -2.4072,  0.5984,     nan,               nan],
          [    nan,     nan,  0.7890,     nan,     nan,     nan,     nan,            0.3440]]]])
xee  tensor([[[[     nan,      nan,      nan,  13.8023,  13.8023,      nan,  13.8023,            13.8023],
          [     nan,      nan,      nan,      nan, -13.8155,      nan,  13.8023,            13.8023]],

         [[     nan,      nan,      nan,      nan, -13.8155,   0.3987,  13.8023,            13.8023],
          [     nan,      nan,   1.3189,      nan,      nan,  13.8023,  13.8023,            -0.6454]]]])
x_c  tensor([[[[     nan,      nan,      nan,  13.8023,  13.8023,      nan,      nan,                nan],
          [     nan,      nan,      nan,  13.8023, -13.8155,  13.8023,  13.8023,            13.8023]],

         [[     nan,      nan,      nan,      nan, -13.8155,   0.3987,      nan,                nan],
          [     nan,      nan,   1.3189,  13.8023,  13.8023,  13.8023,  13.8023,            -0.6454]]]])
```

As you can see in the non-contiguous case the nan is not correctly propagated

cc @VitalyFedyunin @jamesr66a @ppwwyyxx @mruberry @walterddr

**Comment by Flamefire** [comment]
I debugged the issue a bit and the problem is, that `min(x, y)` is not really defined when either argument is NaN. I found that GCC 6.3 uses simple comparisons, while GCC 7+ uses intrinsics on POWER. The difference can be seen at https://godbolt.org/z/cbq71c

This yields a hardware instruction for `min(x,y)` which given NaNs doesn't consistently return NaN or the finite value. Maybe with an effect of vectorization.

In this case there is a `clamp` defined as `c[i] = a[i] < min_vec[i] ? min_vec[i] : (a[i] > max_vec[i] ? max_vec[i] : a[i]);` where the parenthesized part is replaced by the hardware instruction.

Not sure how to solve this, maybe simply say that NaN must be avoided. In this case the test needs fixing.

**Comment by ejguan** [comment]
Even if I can not reproduce the test failure, I think the tests should to be fixed. All in-place operators will modify `x`, which is not our expectation. (`x_c` won't change the memory of `x`, since `x` is not in `contiguous_format`)

**Comment by Flamefire** [comment]
Seemingly this depends on register allocation and which operations are used. I.e. `clamp(x,l,h) = max(l, min(x,h))` and the compiler may choose different hardware instructions for this min/max. And those have different behavior related to NaN especially differing for signaling and quite NaN and which operand is first/last. See e.g. https://ibm.ent.box.com/s/1hzcwkwf8rbju5h9iyf44wm94amnlcrv p. 578/588

**Comment by VitalyFedyunin** [comment]
Interesting, so far we are not doing any optimizations for Power architecture. That means something goes funny with compiler optimization here https://github.com/pytorch/pytorch/blob/af8c75e2116a03aeaef131ef8007fae05546a3a3/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp#L144

Can you please try to compile all code with -O0 to see if issue still remains?

**Comment by Flamefire** [comment]
Building with `DEBUG=1` is a bit tricky due to an error
> test_api/CMakeFiles/test_api.dir/static.cpp.o:static.cpp:vtable for A: error: relocation refers to global symbol "A::~A()", which is defined in a discarded section

When disabling the build of that I was able to confirm the issue does not exists at `-O0`. This is expected. When not optimizing the check used is a pretty much literal `if(a < b) use a; else use b;` which is always false (and hence uses b) when any of a/b is NaN. With optimizations this is replaced (maybe) by a single `min(a,b)` instruction which may define the behavior differently when NaN is encountered and may even differ between different kinds of NaNs.

With some lengthy hours of debugging I found the issue: I injected printfs after the clamp operation and noticed that the nan propagation alternates: (Only) for every 2nd Vec256 they are propagated. Checking the assembly I see an unrolled loop on 2 levels: The first Vec256 has 8 `xsmincdp vs24,vs25,vs24` instructions and the 2nd Vec256 has 8 `xsmincdp vs24,vs24,vs25` (format is target, src1, src2)
And that operation is defined to return src2 if any src* is NaN. So GCCs register allocation decided to reuse the other register for the result in the 2nd unrolled iteration. To me this looks like a compiler bug as that translation is invalid according to NaN handling rules. Using GCC 8.3.0 here, for reference.
